ostd/src/io_mem.rs

```diff
@@ -2,27 +2,43 @@
 
 //! I/O memory.
 
-use core::ops::Range;
+use core::ops::{Deref, Range};
+
+use align_ext::AlignExt;
+use cfg_if::cfg_if;
 
 use crate::{
     mm::{
-        kspace::LINEAR_MAPPING_BASE_VADDR, paddr_to_vaddr, FallibleVmRead, FallibleVmWrite,
-        HasPaddr, Infallible, Paddr, PodOnce, Vaddr, VmIo, VmIoOnce, VmReader, VmWriter,
+        kspace::kva::{Kva, UnTracked},
+        page_prop::{CachePolicy, PageFlags, PageProperty, PrivilegedPageFlags},
+        HasPaddr, Paddr, PodOnce, VmIo, VmIoOnce, VmReader, VmWriter, PAGE_SIZE,
     },
     prelude::*,
-    Error,
 };
 
+cfg_if! {
+    if #[cfg(feature = "cvm_guest")] {
+        use tdx_guest::tdx_is_enabled;
+    }
+}
+
 /// I/O memory.
 #[derive(Debug, Clone)]
 pub struct IoMem {
-    virtual_address: Vaddr,
+    kva: Arc<Kva<UnTracked>>,
+    // The actual mapped range is `kva.start + offset..kva.start + offset + limit`
+    offset: usize,
     limit: usize,
 }
 
 impl HasPaddr for IoMem {
     fn paddr(&self) -> Paddr {
-        self.virtual_address - LINEAR_MAPPING_BASE_VADDR
+        self.kva
+            .deref()
+            .get_untracked_page(self.kva.deref().start())
+            .unwrap()
+            .0
+            + self.offset
```

Comment:
Better not to query over the page table every time when `paddr()` is called, which tries to acquire the lock and may lead to contention. Cache the physical address in the `IoMem` structure.